### ISSUE TITLE
UE 4.16.1 (Plugin v2.1) Diffs problem

### ISSUE THREAD

**Issue by rkorovin** [SRombauts/UE4GitPlugin#44]
Hello.

I recently migrated from Bitbucket to GitHub repo. I use LFS for uasset. Source control plugin version 2.1. My Project has submodule in Content directory. 

Testcase:

1. Select asset->SourceControl->History
1. Click on any commit->Diff against previous revision

Message _"Unable to load assets to diff. Content may no longer be supported?"_. 
Sometimes editor may freeze and don't respond (without Message). 

Sorry for my english. English is not my native language(

**Comment by SRombauts** [comment]
Hi, I didn't test LFS on 4.16 nor on the standalone plugin. I am unable to test this now, but can you:

1. either
    1. Post a complete Log of a short session showing the startup and then the failing operation 
    2. Or at least tell me the results of 'git version' and 'git lfs version' you should use very recent git if possible, at least 2.10.windows
2. Then, remove the "Git (dev) " plugin 2.1 and test again with the "Git (beta version)" plugin 1.0 that I also maintain in Unreal Engine

**Comment by SRombauts** [comment]
Also, I didn't realized you said Content was on a submodule... I am totally unsure about this, it is not tested /officially supported (I should think about this, but lacks the time to)

**Comment by rkorovin** [comment]
'git version' - version 2.11.0.windows.1
'git lfs version' - git-lfs/2.0.2 (GitHub; windows 386; go 1.8; git 85e2aec4)

Test case with message "Unable to load assets to diff. Content may no longer be supported?". At the end I open plugin settings in editor.
[log.txt](https://github.com/SRombauts/UE4GitPlugin/files/1052643/log.txt)

Test case with freeze.
[log2.txt](https://github.com/SRombauts/UE4GitPlugin/files/1052656/log2.txt)

**Comment by rkorovin** [comment]
> Then, remove the "Git (dev) " plugin 2.1 and test again with the "Git (beta version)" plugin 1.0 that I also maintain in Unreal Engine

After that, the editor stopped freezing. But instead of diffs I always see a message "Unable to load assets to diff. Content may no longer be supported?".

**Comment by rkorovin** [comment]
Maybe this will help... Use SourceTree with built-in git & lfs. System variables in win 10 are configured correctly. Since I checked git/lfs versions in SourceTree's bash.

**Comment by SRombauts** [comment]
Hi @rkorovin, thanks a lot for providing these logs and requested information. I cannot see anything suspect in it, so I'll have to dive into this.

As stated before, I currently don't have much spare time (if any) but I'll try to do at least a few tests this evening.
I could/should also add some logs around the method in charge to "Dump content of a specific revision of a file": It uses a special process to do that, and because of that I am blind about it your case.

The crash in the standalone plugin scares me, as I did some "cleanup" of this exact function, but I cannot see anything special on the new code...

**Comment by SRombauts** [comment]
Also, I strongly suspect that the first culprit is that the plugin is picking the SourceTree Git.exe but somehow calling it out of bash break something.

If you don't want to wait for a 2.1.1 of the plugin with more logs or an hypothetical fix, you could also [install the standalone Git for windows+lfs](https://github.com/git-for-windows/git/releases) as I am pretty this could solve your problem.

Cheers!

**Comment by rkorovin** [comment]
> If you don't want to wait for a 2.1.1 of the plugin with more logs or an hypothetical fix, you could also install the standalone Git for windows+lfs as I am pretty this could solve your problem.

I was just thinking about trying to do this. Also the problem can be covered in migration from Bitbucket to Github. Perhaps information about the old remote is stored somewhere. For the purity of the experiment, you can use in tests my project "GameMechanics" on github.

**Comment by rkorovin** [comment]
> If you don't want to wait for a 2.1.1 of the plugin with more logs or an hypothetical fix, you could also install the standalone Git for windows+lfs as I am pretty this could solve your problem.

I do it. It works beter. Now i can see some diffs for 10-15% of files (In other cases I see the same message). I think this is because they are somehow connected with a submodule. For example, actors use ActorComponents implemented in a submodule. And this is not enough for normal work. Because major part of the logic is in the submodule or is connected with it(

**Comment by SRombauts** [comment]
Yes, git submodules are not supported by the plugin (Git LFS or not, I don't think it matters). I could try to improve on this.

I've downloaded your Github project, but it's big and took forever, now it is downloading a plugin from the marketplace.
I'll give it a try latter.

**Comment by rkorovin** [comment]
I, meanwhile, will try to check what will happen if the components of the submodule are implemented on C++. I do not think it will change the situation with related uassets. But at least Diffs can be watched in SourceTree.

**Comment by rkorovin** [comment]
So i test Plugin with c++ Game Module as Git Submodule and it works! I think the problem lies in the fact that submodule content is in the Large File Storage.

**Comment by adnoh** [comment]
@SRombauts  I have the problem still in current engine builds with LFS and without (the "beta" module included in UE).
As soon as my Content like BPs is in a submodule I can't diff it in the Editor anymore which the Message 
**Unable to load assets to diff. Content may no longer be supported?**
This is really anoying - I work a lot with bundling Content in Git Submodules (UE Plugins / DLCs)

**Comment by Bazard** [comment]
Hi there, do you have any plan on adding submodules support in your plugin ? 

I am in a similar situation as @adnoh, I often work with submodules for plugins and DLCs in my projects, it would be really great to be able to use your plugin with them (at least the diff feature). 
Currently, on submodules files we can only see when the pointer in main repo has changed in history, but we can't see the history inside the submodule itself, or doing diff on submodule files.

**Comment by Louspirit** [comment]
This feature is **critical.** As there is no way to make Blueprint libraries, the best way I found to share blueprints between projects is with git submodules. But I can't work efficiently without the git plugin operations.

And there ain't any solutions external to UE for blueprint diff am I right?

**Comment by Louspirit** [comment]
@SRombauts did you take a look at this? Can we help somehow?

**Comment by EvoPulseGaming** [comment]
Same issue here, cannot diff in ue4 on content in submodules as it does not see the history of the submodule repos.
The plugin should be looking in .git/modules/ folder for additional repos to load history on.

in my case:
.git/modules/plugins/x
.git/modules/plugins/y
.git/modules/plugins/z

etc

Basically recursively search the directory, and back out once you find the config file. Then go to the next folder.
